app/controllers/workflows_controller.rb

```diff
@@ -58,29 +56,33 @@ def stage # rubocop:todo Metrics/CyclomaticComplexity
     # else actually execute the task.
     unless params[:next_stage].nil?
       eager_loading = @task.included_for_do_task
-      @batch ||= Batch.includes(eager_loading).find(params[:batch_id])
+      @batch = Batch.includes(eager_loading).find(params[:batch_id])
 
-      unless @batch.editable?
-        flash[:error] = 'You cannot make changes to a completed batch.'
-        redirect_back fallback_location: root_path
+      editable, message = @task.can_process?(@batch, from_previous: true)
+
+      unless editable
+        redirect_back fallback_location: batch_path(@batch), alert: message
         return false
       end
 
       ActiveRecord::Base.transaction do
-        if @task.do_task(self, params)
+        task_success, task_message = @task.do_task(self, params)
+        if task_success
```

Comment:
In order to stop tasks needing to interact with the flash, they now return a combination of a success, and a message.